src/Neo/Cryptography/Crypto.cs

```diff
@@ -112,6 +108,107 @@ public static byte[] Sign(byte[] message, byte[] priKey, ECC.ECCurve ecCurve = n
             return ecdsa.SignData(message, hashAlg);
         }
 
+        /// <summary>
+        /// ECRecover
+        /// </summary>
+        /// <param name="curve">Curve</param>
+        /// <param name="signature">Signature</param>
+        /// <param name="hash">Message hash</param>
+        /// <param name="format">Signature format</param>
+        /// <returns>Allowed Public keys</returns>
+        public static ECC.ECPoint[] ECRecover(ECC.ECCurve curve, byte[] signature, byte[] hash, SignatureFormat format = SignatureFormat.Der)
+        {
+            BigInteger r, s;
+            int recId = 0, recIdTo = 4;
+
+            // Decode signature
+
+            switch (format)
+            {
+                case SignatureFormat.Der:
+                    {
+                        var derSequence = (DerSequence)Asn1Object.FromByteArray(signature);
+                        r = ((DerInteger)derSequence[0]).Value;
+                        s = ((DerInteger)derSequence[1]).Value;
+
+                        if (derSequence.Count == 3)
+                        {
+                            recId = ((DerInteger)derSequence[2]).IntValueExact;
+                            recIdTo = recId + 1;
+                        }
+                        break;
+                    }
+                case SignatureFormat.Fixed32:
+                    {
+                        r = new(1, signature, 0, 32);
+                        s = new(1, signature, 32, 32);
+
+                        if (signature.Length == RecuperableSignatureLength)
+                        {
+                            recId = signature[SignatureLength];
+                            recIdTo = recId + 1;
+                        }
+                        break;
+                    }
+                default: throw new InvalidOperationException("Invalid signature format");
+            }
+
+            // Validate values
+
+            if (recId < 0 || recId >= 4) throw new ArgumentException("v should be positive less than 4");
+            if (r.SignValue < 0) throw new ArgumentException("r should be positive");
+            if (s.SignValue < 0) throw new ArgumentException("s should be positive");
```

Comment:
https://github.com/Hecate2/ECrecover/blob/c3ec9b4ca4f74e66c426b48452f86cf643a9f6fb/Program.cs#L83-L84
```csharp
        if (r.CompareTo(BigInteger.One) < 0 || r.CompareTo(curve.N) > 0) throw new ArgumentException($"Invalid r value {r}; expected [1, {curve.N}]");
        if (s.CompareTo(BigInteger.One) < 0 || s.CompareTo(curve.N) > 0) throw new ArgumentException($"Invalid s value {s}; expected [1, {curve.N}]");
```

Reply:
Optionally we can require s < n / 2 if v is not specified, for Ethereum